build-system/babel-plugins/babel-plugin-transform-log-methods/index.js

```diff
@@ -0,0 +1,288 @@
+const defaultMessagesPath = messagesPath;
+
+/**
+ * Approximate length of a compressed message expansion call to determine
+ * whether an instance of message indirection actually reduces binary size.
+ */
+const roughMinifiedExpansionLength = '["xx"]'.length;
+
+const assertAliases = singletonFunctions.map(prefix => `${prefix}Assert`);
+
+/**
+ * Determines a filepath relative to the repo root.
+ * @param {string} path
+ * @return {string}
+ */
+const relativeToRoot = path => `${__dirname}/../../../${path}`;
+
+/**
+ * Reads the messages table from the JSON file.
+ * @param {string} messagesPath
+ * @return {!Object<string, string>}
+ */
+function getMessages(messagesPath) {
+  try {
+    return JSON.parse(fs.readFileSync(messagesPath));
+  } catch {
+    // When non-existent or empty just return an empty object. Transformations
+    // that read will write as well.
+    return {};
+  }
+}
+
+/**
+ * Writes the messages table to the JSON file.
+ * @param {string} messagesPath
+ * @param {!Object<string, string>} obj
+ */
+function writeMessages(messagesPath, obj) {
+  const json = JSON.stringify(obj, /* replacer */ null, /* spaces */ 2);
+  const messagesPathDir = path.dirname(messagesPath);
+  if (!fs.existsSync(messagesPathDir)) {
+    fs.mkdirSync(messagesPathDir);
+  }
+  fs.writeFileSync(messagesPath, json);
+}
+
+/**
+ * Gets a message id from the table, or adds a new entry for a message if
+ * non-existent and returns its new id.
+ * @param {Object<string, string>} messages
+ * @param {string} message
+ * @param {function():number} getMessageId
+ * @return {string} Short message id.
+ */
+function getOrCreateShortMessageId(messages, message, getMessageId) {
+  if (messages[message]) {
+    return messages[message];
+  }
+  // Base-62 radix for best utilization of ascii space.
+  const shortMessageId = base62.encode(getMessageId());
+  messages[message] = shortMessageId;
+  return shortMessageId;
+}
+
+/**
+ * Builds a message template using printf syntax from a starting node.
+ * @param {@babel/types} t
+ * @param {!Node} node
+ * @param {!Array<!Node>} interpolationArgs
+ * @return {string} Template for the message.
+ */
+function buildMessage(t, node, interpolationArgs) {
+  if (t.isStringLiteral(node)) {
+    return node.value;
+  }
+
+  if (t.isBinaryExpression(node, {operator: '+'})) {
+    return (
+      buildMessage(t, node.left, interpolationArgs) +
+      buildMessage(t, node.right, interpolationArgs)
+    );
+  }
+
+  if (t.isTemplateLiteral(node)) {
+    let quasied = '';
+    let i = 0;
+    for (; i < node.quasis.length - 1; i++) {
+      quasied += node.quasis[i].value.cooked;
+      quasied += buildMessage(t, node.expressions[i], interpolationArgs);
+    }
+    quasied += node.quasis[i].value.cooked;
+    return quasied;
+  }
+
+  interpolationArgs.push(node);
+  return '%s';
+}
+
+/**
+ * @param {@babel/types} t
+ * @param {!Node} node
+ * @return {?Object}
+ */
+function getTransformableCallExpressionMeta(t, {callee}) {
```

Comment:
Nit: Anything that relies on `t` should be done inside the `module.export` function to make call sites a little cleaner. Declarations will still only be evaluated once.

Reply:
Hm. Inlined, but now there's deep indentation and `if-else-return` which arguably is less readable. Up to you.